lib/selection.js

```diff
@@ -0,0 +1,202 @@
+/** A selection of pieces to download.
+ * @typedef {Object} SelectionItem
+ * @property {number} from
+ * @property {number} to
+ * @property {number} offset
+ * @property {number} priority
+ * @property {function} notify
+ * @property {boolean} isStreamSelection
+ */
+
+export class Selection {
+  /**
+   * @type {Array<SelectionItem>}
+   * @private
+   */
+  _items = []
+
+  /**
+   * @public
+   * @param {{from: number, to: number}} item Interval to be removed from the selection
+   * @param {boolean} [isStreamSelection]
+   */
+  remove (item, isStreamSelection = false) {
+    for (let i = 0; i < this._items.length; i++) {
+      const existing = this._items[i]
+
+      if (existing.isStreamSelection) {
+        if (isStreamSelection) {
+          // If both are stream selections and they match, then we remove the first matching item, then we break the loop
+          if (existing.from === item.from && existing.to === item.to) {
+            this._items.splice(i, 1)
+            // for stream selections, we only remove one item at a time
+            // ergo we break the loop after removing the first matching item
+            break
+          }
+        } else {
+          // we only remove stream selections when the `isStreamSelection` flag is true and they match
+          continue
+        }
+      }
+
+      if (isLowerIntersecting(item, existing)) {
+        existing.to = item.from - 1
+        continue
+      }
+      if (isUpperIntersecting(item, existing)) {
+        existing.from = item.to + 1
+        continue
+      }
+      if (isInsideExisting(item, existing)) {
+        const existingStart = { ...existing, to: item.from - 1 }
+        const existingEnd = { ...existing, from: item.to + 1 }
+        this._items.splice(i, 1, existingStart, existingEnd)
+        continue
+      }
+      if (isCoveringExisting(item, existing)) {
+        this._items.splice(i, 1)
+        i -= 1
+        continue
+      }
+    }
+  }
+
+  /**
+   * @public
+   * @param {SelectionItem} newItem
+   * @param {boolean} [isStreamSelection]
+   */
+  insert (newItem, isStreamSelection = false) {
+    if (newItem.from > newItem.to) {
+      throw new Error('Invalid interval')
+    }
+    if (!isStreamSelection) {
+      this.remove(newItem)
+    }
+    this._items.push({ ...newItem, isStreamSelection })
+  }
+
+  /**
+   * @public
+   * @param {(a: SelectionItem, b: SelectionItem) => number} sortFn
+   */
+  sort (sortFn = (a, b) => a.from - b.from) {
+    this._items.sort(sortFn)
+  }
+
+  /**
+   * @public
+   */
+  isEmpty () {
```

Comment:
this is VERY redundant

Reply:
can't really argue with that haha 😄 removed it.